File: lib/vec2.lua
--- @class Vec2
--- @field x integer
--- @field y integer

Vec2 = {}

Vec2.__index = Vec2

Vec2.__tostring = function(self)
  return string.format("Vec2(%d, %d)", self.x, self.y)
end

Vec2.__eq = function(self, other)
  return self.x == other.x and self.y == other.y
end

Vec2.__add = function(self, other)
  return Vec2.new(self.x + other.x, self.y + other.y)
end

Vec2.__sub = function(self, other)
  return Vec2.new(self.x - other.x, self.y - other.y)
end

function Vec2.new(x, y)
  local self = setmetatable({}, Vec2)
  self.x = x or 0
  self.y = y or 0
  return self
end

function Vec2:tohash()
  return tostring(self)
end

function Vec2:copy()
  return Vec2.new(self.x, self.y)
end

return Vec2

File: lib/vec2.ts
import type { StringConvertible } from "./string-convertible.js";

export class Vec2 implements StringConvertible {
  constructor(public x: number, public y: number) {}

  equals(other: Vec2): boolean {
    return this.x == other.x && this.y == other.y;
  }

  add(other: Vec2): Vec2 {
    return new Vec2(this.x + other.x, this.y + other.y);
  }

  subtract(other: Vec2): Vec2 {
    return new Vec2(this.x - other.x, this.y - other.y);
  }

  toString(): string {
    return `${Vec2.name}(${this.x}, ${this.y})`;
  }
}
